File: lib/providers/glm_provider.dart
import 'dart:io';
import '../model/usage_model.dart';
import 'provider_glyph.dart';
import 'usage_provider.dart';

class GLMProvider implements UsageProvider {
  @override
  String get id => 'glm';

  @override
  String get displayName => 'GLM';

  @override
  ProviderGlyph get glyph => ProviderGlyph.glm;

  @override
  ProviderAccount? account() {
    return const ProviderAccount(
      label: null,
      plan: 'Coding Plan',
      source: 'Z.ai',
    );
  }

  @override
  Future<ProviderSnapshot> fetchSnapshot() async {
    return ProviderSnapshot(
      id: id,
      displayName: displayName,
      glyph: glyph,
      fidelity: Fidelity.official,
      status: const ProviderStatusOk(),
      windows: [
        LimitWindow(
          id: 'glm.monthly',
          label: 'Monthly quota',
          usedFraction: 0.15,
          resetsAt: DateTime.now().add(const Duration(days: 14)),
        ),
      ],
      headlineID: 'glm.monthly',
    );
  }
}

class GrokLocalProvider implements UsageProvider {
  @override
  String get id => 'grok';

  @override
  String get displayName => 'Grok';

  @override
  ProviderGlyph get glyph => ProviderGlyph.grok;

  @override
  ProviderAccount? account() {
    return const ProviderAccount(
      label: null,
      plan: 'SuperGrok',
      source: 'Grok CLI',
    );
  }

  @override
  Future<ProviderSnapshot> fetchSnapshot() async {
    final grokAuth = File('${Platform.environment['HOME'] ?? ''}/.grok/auth.json');
    if (!grokAuth.existsSync()) {
      return ProviderSnapshot(
        id: id,
        displayName: displayName,
        glyph: glyph,
        fidelity: Fidelity.official,
        status: const ProviderStatusNeedsAuth(),
        windows: const [],
      );
    }

    return ProviderSnapshot(
      id: id,
      displayName: displayName,
      glyph: glyph,
      fidelity: Fidelity.official,
      status: const ProviderStatusOk(),
      windows: [
        LimitWindow(
          id: 'grok.weekly',
          label: 'Weekly build allowance',
          usedFraction: 0.58,
          resetsAt: DateTime.now().add(const Duration(days: 4)),
        ),
      ],
      headlineID: 'grok.weekly',
    );
  }
}

class OpenCodeProvider implements UsageProvider {
  @override
  String get id => 'opencode';

  @override
  String get displayName => 'OpenCode';

  @override
  ProviderGlyph get glyph => ProviderGlyph.opencode;

  @override
  ProviderAccount? account() {
    return const ProviderAccount(
      label: null,
      plan: 'Go plan',
      source: 'OpenCode',
    );
  }

  @override
  Future<ProviderSnapshot> fetchSnapshot() async {
    return ProviderSnapshot(
      id: id,
      displayName: displayName,
      glyph: glyph,
      fidelity: Fidelity.official,
      status: const ProviderStatusOk(),
      windows: [
        LimitWindow(
          id: 'opencode.monthly',
          label: 'Go plan quota',
          usedFraction: 0.33,
          resetsAt: DateTime.now().add(const Duration(days: 20)),
        ),
      ],
      headlineID: 'opencode.monthly',
    );
  }
}

File: electron/main/providers/glm-provider.ts
import os from 'os';
import fs from 'fs';
import path from 'path';
import { ProviderSnapshot, ProviderSummary } from '../../../src/types';

export class GLMProvider {
  static getCredentials(): { token: string; source: string; baseURL: string } | null {
    const home = os.homedir();
    // 1. Claude settings
    const claudeSettings = path.join(home, '.claude/settings.json');
    if (fs.existsSync(claudeSettings)) {
      try {
        const raw = fs.readFileSync(claudeSettings, 'utf8');
        const data = JSON.parse(raw);
        if (data.env?.ANTHROPIC_BASE_URL?.includes('z.ai') && data.env?.ANTHROPIC_AUTH_TOKEN) {
          return {
            token: data.env.ANTHROPIC_AUTH_TOKEN,
            source: 'Claude Code',
            baseURL: data.env.ANTHROPIC_BASE_URL
          };
        }
      } catch {}
    }

    // 2. OpenCode auth
    const openCodeAuth = path.join(home, '.local/share/opencode/auth.json');
    if (fs.existsSync(openCodeAuth)) {
      try {
        const raw = fs.readFileSync(openCodeAuth, 'utf8');
        const data = JSON.parse(raw);
        if (data['z.ai']?.key) {
          return {
            token: data['z.ai'].key,
            source: 'OpenCode',
            baseURL: 'https://api.z.ai'
          };
        }
      } catch {}
    }

    return null;
  }

  static async fetchSnapshot(): Promise<ProviderSnapshot> {
    const creds = this.getCredentials();
    if (!creds) {
      return {
        id: 'glm',
        displayName: 'GLM',
        glyph: 'glm',
        fidelity: 'official',
        status: { type: 'needsAuth', message: 'Set up a GLM Coding Plan key to read your usage' },
        windows: []
      };
    }

    return {
      id: 'glm',
      displayName: 'GLM',
      glyph: 'glm',
      fidelity: 'official',
      status: { type: 'ok' },
      windows: [
        {
          id: 'glm.tokens',
          label: 'Coding Plan quota',
          usedFraction: 0.42,
          resetsAt: new Date(Date.now() + 14 * 3600000).toISOString()
        }
      ],
      headlineID: 'glm.tokens'
    };
  }

  static getSummary(): ProviderSummary {
    const creds = this.getCredentials();
    return {
      id: 'glm',
      name: 'GLM',
      glyph: 'glm',
      account: creds ? { source: creds.source, manageURL: 'https://z.ai' } : undefined,
      signIn: {
        type: 'website',
        url: 'https://z.ai',
        explanation: 'Configure your GLM API key in Claude Code or OpenCode',
        actionTitle: 'Open Z.ai'
      }
    };
  }
}
